_blog/2024-01-23-using-serde-rust.mdx

```diff
@@ -0,0 +1,341 @@
+pub struct MyStruct {
+    my_message: String
+}
+```
+
+Now when you serialize this struct, `my_message` should automatically turn into `myMessage`! Perfect for working with APIs written in other languages or with different conventions.
+
+If you’d prefer to not wrap fields in `Option`, you can also implement default values by using `#[serde(default)]`. This simply allows fields to be filled in with default values instead of automatically erroring out. You can also use `#[serde(default = "path")]` to be able to point to functions for providing the automatic default. For example, this struct and function:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+pub struct MyStruct {
+	#[serde(path = "my_function")]
+    my_message: String,
+}
+
+fn my_function() -> String {
+    "Hello world!".to_string()
+}
+```
+
+`serde` also offers other useful attributes, like being able to deny unknown fields using `#[serde(deny_unknown_fields)]` on top of the struct. This allows you to make sure that the struct is exactly as-is when serializing and deserializing. 
+
+### Deserializing and Serializing enums
+
+Let’s examine this enum type:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+enum MyEnum {
+	Data { id: String, data: Value },
+    SomeOtherData { id: i32, name: String }
+}
+```
+
+Note that when converting to and from this enumstruct, it can take two options:
+
+- A String field named “id” and a JSON value with the key “data” (this can be a map, a value or anything that the `Json` value can hold)
+- An `i32` field named “id” and a `String` field named “name”
+
+You can then match the enum variant for further processing. 
+
+When the first enum variant is written in JSON, you can see that it should correspond with this:
+
+```json
+{
+    "Data": {
+        "id": "your_id_here",
+        "data": { .. }
+    }
+}
+```
+
+This type of data is “externally tagged” - meaning the data is characterized by the identifier being on the outside of the JSON object. We can add inline tagging so that the identifier is on the inside of the crate - let’s have a look at what this would look like:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+#[serde(tag = "type")]
+enum MyEnum {
+	Data { id: String, data: Value },
+    SomeOtherData { id: i32, name: String }
+}
+```
+
+Now the JSON representation looks like this:
+
+```json
+{
+    "type": "Data",
+    "id": "your_id_here",
+    "data": { .. }
+}
+```
+
+Interested in reading more? The `serde` documentation has a page on tagging which you can find [here.](https://serde.rs/enum-representations.html)
+
+## Crates that work well with Serde
+
+### serde_with
+
+`serde_with` is a crate that provides custom de/serialization helpers to use with `serde`'s `with` annotation. It works by replacing the `with` annotation with a new one called `serde_as`. With this new attribute macro you can do quite a few things:
```

Comment:
It might have been interesting to introduce the `with` annotation first, otherwise someone new won't be able to understand what it does.